### ISSUE TITLE
Behaviour with average=`sample`, mdmc_average=`samplewise` seems confusing.

### ISSUE THREAD

**Issue by AndreiSiliUiPath** [Lightning-AI/metrics#452]
## 🐛 Bug

Computing F1 with `average="samples"` and `mdmc_average=samplewise` seems to take into consideration all samples, including those that are ignored.

Not sure if this is a bug or I am simply confused by what average="samples" means or the docs are not clear.

#### Code sample

```
import torch
import torchmetrics

>>> pred = torch.tensor([[0,1,2,3]])
>>> target = torch.tensor([[0,1,2,3]])
>>> pred.shape
torch.Size([1, 4])
>>> target.shape
torch.Size([1, 4])
>>> torchmetrics.functional.f1(pred, target, ignore_index=0, average="samples", mdmc_average="samplewise")
tensor(0.7500)
>>> torchmetrics.functional.f1(pred, target, ignore_index=0, average="micro", mdmc_average="samplewise")
tensor(1.)
```

### Expected behavior

I am not sure what the expected behaviour is and what the utility of `average="samples"` is. In this case, I want to take the average over the samples first, excluding samples which I marked as ignored in the target. Then, I want to take the average over the batch. I see now that `average="micro"` and `mdmc_average="samplewise"` achieves just this, but my first impression was that this should be achieved with `average="samples"`.

### Environment

- PyTorch Version (e.g., 1.0): 1.8.1
- OS (e.g., Linux): Linux Ubuntu 18.04
- How you installed PyTorch (`conda`, `pip`, source): pip
- Build command you used (if compiling from source):
- Python version: 3.8.10
- CUDA/cuDNN version: 11.1

### Additional context

I have run into this problem for a language modelling task with a transformer-based network. It is common practice there to have padding marked as `-1` and out-of-vocabulary tokens as `0`.

**Comment by MeganClark-Robinson** [comment]
What's the difference between 'samplewise' and 'global' when solving a mdmc problem?

**Comment by SkafteNicki** [comment]
@MeganClark-Robinson:
* 'global' is probably the easiest to understand. This just means that all additional dimensions are just considered extra observations. So if you have tensor with provabilities that have shape `[N, C, D]` where C is the number classes it will internally just be reshaped to a `[N*D, C]` tensor.
* `samplewise` means that the score will be calculated per sample and then averaged. Going back to the `[N, C, D]` tensor, we basically run a for loop along the N dimension, calculating the metric on each sub `[C, D]` matrix and then averaging them in the end

**Comment by SkafteNicki** [comment]
@AndreiSiliUiPath thanks for bringing up this issue.
I am not sure at the moment if there is a bug here or the result should be like that. I have to dig a bit deeper into the code to figure it out.

**Comment by AndreiSiliUiPath** [comment]
@SkafteNicki happy to help. I did some digging and I think the issue relates to the interaction between `torchmetrics.functional.classification.f_beta._fbeta_compute` and `torchmetrics.functional.classification.stat_scores._reduce_stat_scores`:
- `_reduce_stat_scores` expects `denom` to be set to -1 for ignore cases. ```
line:179
ignore_mask = denominator < 0
...
line:194:195
if average not in (AverageMethod.MICRO, AverageMethod.NONE, None):
    weights = weights / weights.sum(dim=-1, keepdim=True)
```
- `_fbeta_compute` does not ignore the index for average method "samples".
```
line:62:68
if ignore_index is not None:
    if average not in (AvgMethod.MICRO, AvgMethod.SAMPLES) and mdmc_average == MDMCAverageMethod.SAMPLEWISE:
        num[..., ignore_index] = -1
        denom[..., ignore_index] = -1
    elif average not in (AvgMethod.MICRO, AvgMethod.SAMPLES):
        num[ignore_index, ...] = -1
        denom[ignore_index, ...] = -1
```
- This results in weights being 1 for index 0, because it is not ignored. In the case of average="micro", this is not an issue because values are already averaged over samples when they reach `_fbeta_compute`. Here's the `num` and `denom` values:
```
# average="samples", mdmc_average="samplewise"
>>> num
tensor([[0., 2., 2., 2.]])
>>> denom
tensor([[1., 2., 2., 2.]])
# average="micro", mdmc_average="samplewise"
>>> num
tensor([2.])
>>> denom
tensor([2.])
```

**Comment by Borda** [comment]
@SkafteNicki so this shall be addressed within the suggested classification revamp, correct? :rabbit:

**Comment by SkafteNicki** [comment]
Issue will be fixed by classification refactor: see this issue https://github.com/Lightning-AI/metrics/issues/1001 and this PR https://github.com/Lightning-AI/metrics/pull/1195 for all changes

Small recap: This issue describes how the behaviour of `average="sample"` and `mdmc_average="samplewise"` is confusing and it is not clear what the computed result actually should be. During the refactor it was decided to completely drop `average='samples'` option for now as it usually is an option that only makes sense for multilabel problems not multiclass problems (thats atleast how sklearn do it). If the goal is to calculate an score as an average over some metric calculated per sample this should be possible just through the new `multidim_average="samplewise"` argument (just a renamed version of `mdmc_average`. An example:

```python
from torchmetrics.functional import multiclass_f1_score
import torch

# two samples where everything match except index [0,0]
target = torch.tensor([[0, 1, 2, 3], [1, 1, 2, 3]])
preds = tensor([[1, 1, 2, 3], [1, 1, 2, 3]])

# calculate globally
multilabel_jaccard_index(preds, target, num_labels=3, average="micro")
# tensor(0.8750)

# calculate per sample
multilabel_jaccard_index(preds, target, num_labels=3, average="micro", multidim_average)
# tensor([0.7500, 1.0000])

# calculate per sample, ignore where target==0
multiclass_f1_score(preds, target, num_classes=4, average='micro', multidim_average='samplewise', ignore_index=0)
# tensor([1.0000, 1.0000])
```
which seems to do the right thing as when we ignore the wrongly classified sample we get an score of 1.0 for each sample.
Issue will be closed when https://github.com/Lightning-AI/metrics/pull/1195 is merged.
